src/crypto/qat/qcccrypto.cc

```diff
@@ -322,150 +327,187 @@ bool QccCrypto::perform_op(unsigned char* out, const unsigned char* in,
-    dout(15) << "Instantiation complete for " << avail_inst << dendl;
+  } else {
+    dout(1) << "Session alloc failed with status = " << status << dendl;
   }
+  return status;
+}
 
-  // Section that runs on every call
-  // Identify the operation and assign to session
-  qcc_sess[avail_inst].sess_stp_data.cipherSetupData.cipherDirection = op_type;
-  qcc_sess[avail_inst].sess_stp_data.cipherSetupData.pCipherKey = (Cpa8U *)key;
-
-  stat = cpaCySymInitSession(qcc_inst->cy_inst_handles[avail_inst],
-      NULL,
-      &(qcc_sess[avail_inst].sess_stp_data),
-      qcc_sess[avail_inst].sess_ctx);
-  if (stat != CPA_STATUS_SUCCESS) {
-    derr << "Unable to init session" << dendl;
-    return false;
-  }
+bool QccCrypto::symPerformOp(int avail_inst,
+                              CpaCySymSessionCtx sessionCtx,
+                              const Cpa8U *pSrc,
+                              Cpa8U *pDst,
+                              Cpa32U size,
+                              Cpa8U *pIv,
+                              Cpa32U ivLen) {
+  CpaStatus status = CPA_STATUS_SUCCESS;
+  Cpa32U one_batch_size = chunk_size * MAX_NUM_SYM_REQ_BATCH;
+  Cpa32U iv_index = 0;
+  for (Cpa32U off = 0; off < size; off += one_batch_size) {
+    CompletionHandle complete;
+    for (Cpa32U offset = off, i = 0; offset < size && i < MAX_NUM_SYM_REQ_BATCH; offset += chunk_size, i++) {
+      CpaCySymDpOpData *pOpData = qcc_op_mem[avail_inst].sym_op_data[i];
+      Cpa8U *pSrcBuffer = qcc_op_mem[avail_inst].src_buff[i];
+      Cpa8U *pIvBuffer = qcc_op_mem[avail_inst].iv_buff[i];
+      Cpa32U process_size = offset + chunk_size <= size ? chunk_size : size - offset;
+      if (CPA_STATUS_SUCCESS == status) {
+        // copy source into buffer
+        memcpy(pSrcBuffer, pSrc + offset, process_size);
+        // copy IV into buffer
+        memcpy(pIvBuffer, &pIv[iv_index * ivLen], ivLen);
+        iv_index++;
+
+        complete.add_one();
+
+        //pOpData assignment
+        pOpData->thisPhys = qaeVirtToPhysNUMA(pOpData);
+        pOpData->instanceHandle = qcc_inst->cy_inst_handles[avail_inst];
+        pOpData->sessionCtx = sessionCtx;
+        pOpData->pCallbackTag = (void *)&complete;
+        pOpData->cryptoStartSrcOffsetInBytes = 0;
+        pOpData->messageLenToCipherInBytes = process_size;
+        pOpData->iv = qaeVirtToPhysNUMA(pIvBuffer);
+        pOpData->pIv = pIvBuffer;
+        pOpData->ivLenInBytes = ivLen;
+        pOpData->srcBuffer = qaeVirtToPhysNUMA(pSrcBuffer);
+        pOpData->srcBufferLen = process_size;
+        pOpData->dstBuffer = qaeVirtToPhysNUMA(pSrcBuffer);
+        pOpData->dstBufferLen = process_size;
+
+        status = cpaCySymDpEnqueueOp(pOpData, CPA_FALSE);
+        if (unlikely(CPA_STATUS_SUCCESS != status)) {
+          dout(1) << "cpaCySymDpEnqueueOp failed. (status = " << status << ")" << dendl;
+          // submit all request, clean the queue
+          cpaCySymDpPerformOpNow(qcc_inst->cy_inst_handles[avail_inst]);
```

Comment:
Is it can clean queue ? If so we should return.

Reply:
yes, add `return false;`